### PR TITLE
Add StableDiffusion repaint pipeline

### PR DESCRIPTION
**Pull request by nathanielherman**

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_repaint.py [hunk 1]
```diff
@@ -190,6 +190,7 @@ def set_timesteps(
 
         timesteps = np.array(timesteps) * (self.config.num_train_timesteps // self.num_inference_steps)
         self.timesteps = torch.from_numpy(timesteps).to(device)
+        self.timesteps += self.config.steps_offset
```
Comment (nathanielherman) [inline src/diffusers/schedulers/scheduling_repaint.py:193]: repaint scheduler wasn't doing this but other schedulers do, I assume this step is supposed to be here? (it doesn't seem to affect output much)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 2]
```diff
@@ -0,0 +1,606 @@
...
+        ):
+            raise ValueError(
+                f"`callback_steps` has to be a positive integer but is {callback_steps} of type"
+                f" {type(callback_steps)}."
+            )
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.StableDiffusionImg2ImgPipeline.get_timesteps
+    def get_timesteps(self, num_inference_steps, strength, device):
+        # get the original timestep using init_timestep
+        # TODO: steps_offset is usually 1, so this effectively cuts the first step out when strength=1.0, is that desired? (for inpaint/img2img)
```
Comment (nathanielherman) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:411]: is this a bug in inpaint_legacy or intended? (ie inpaint_legacy will remove the first step when steps_offset is set to a default of 1)

**Review by patrickvonplaten — COMMENTED** [review]
I'm actually ok to leave it here given that the code uses a lot of "Copied from statements" - @anton-l what do you think?

**Comment by nathanielherman** [comment]
Bump on this PR! Also @patrickvonplaten wdym by "leave it here"?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+        unet ([`UNet2DConditionModel`]): Conditional U-Net architecture to denoise the encoded image latents.
+        scheduler ([`SchedulerMixin`]):
+            A scheduler to be used in combination with `unet` to denoise the encoded image latents. Can be one of
+            [`DDIMScheduler`], [`LMSDiscreteScheduler`], or [`PNDMScheduler`].
+        safety_checker ([`StableDiffusionSafetyChecker`]):
+            Classification module that estimates whether generated images could be considered offensive or harmful.
+            Please, refer to the [model card](https://huggingface.co/runwayml/stable-diffusion-v1-5) for details.
+        feature_extractor ([`CLIPFeatureExtractor`]):
+            Model that extracts features from generated images to be used as inputs for the `safety_checker`.
+    """
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:93]:
```suggestion
    """
    _optional_components = ["safety_checker", "feature_extractor"]
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+    w, h = image.size
+    w, h = map(lambda x: x - x % 32, (w, h))  # resize to integer multiple of 32
+    image = image.resize((w, h), resample=PIL.Image.LANCZOS)
+    image = np.array(image).astype(np.float32) / 255.0
+    image = image[None].transpose(0, 3, 1, 2)
+    image = torch.from_numpy(image)
+    return 2.0 * image - 1.0
+
+
+def preprocess_mask(mask):
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:54]:
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_inpaint_legacy.preprocess_mask
def preprocess_mask(mask):
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+        guidance_scale: Optional[float] = 7.5,
+        negative_prompt: Optional[Union[str, List[str]]] = None,
+        num_images_per_prompt: Optional[int] = 1,
+        eta: Optional[float] = 0.0,
+        generator: Optional[torch.Generator] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, torch.FloatTensor], None]] = None,
+        callback_steps: Optional[int] = 1,
+        **kwargs,
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:456]:
```suggestion
```

**Review by anton-l — COMMENTED** [review]
Hi @nathanielherman, the PR looks great already! We've had to make some changes to the Stable Diffusion pipelines last week to accommodate for SD 2.0, so we'll need to do some tweaking here as well, hope that's ok :)

Most of the updates will be copied over when you run `python utils/check_copies.py --fix_and_overwrite` thanks to the `# Copied from` comments!

Also, could you add one integration test similar to https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint_legacy.py#L352 so that we have the same reference? :pray:

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+)
+from ...utils import deprecate, logging
+from . import StableDiffusionPipelineOutput
+from .safety_checker import StableDiffusionSafetyChecker
+
+
+logger = logging.get_logger(__name__)
+
+
+def preprocess_image(image):
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:44]:
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_inpaint_legacy.preprocess_image
def preprocess_image(image):
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+        latents = torch.cat([noise * self.scheduler.init_noise_sigma] * batch_size * num_images_per_prompt, dim=0)
+        return latents, init_latents_orig
+
+    @torch.no_grad()
+    def __call__(
+        self,
+        prompt: Union[str, List[str]],
+        init_image: Union[torch.FloatTensor, PIL.Image.Image],
+        mask_image: Union[torch.FloatTensor, PIL.Image.Image],
+        num_inference_steps: Optional[int] = 50,
```
Comment (Randolph-zeng) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:444]: Sorry to jump in, but strength argument is missing in here

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+        mask = torch.cat([mask] * batch_size * num_images_per_prompt)
+
+        t_last = timesteps[0] + 1
+        # 9. Denoising loop
+        for i, t in enumerate(self.progress_bar(timesteps)):
+            # expand the latents if we are doing classifier free guidance
+            latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
+            latent_model_input = self.scheduler.scale_model_input(latent_model_input, t)
+
+            if t >= t_last:
```
Comment (Randolph-zeng) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:570]: and sorry to jump in again, but this line actually causes a bug:
in original repaint pipeline, the t >= last is actually a condition that wraps the main denoise logic, if you instead put such condition check in here, the first time that t >= last is satisfied, the latents size will be doubled but skipped the unet forward that puts its shape back, thus causing error in the next round.
An easy way to reproduce this is to set jump_n_sample to 2 or anything larger than 1

**Review by Randolph-zeng — CHANGES_REQUESTED** [review]

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py (hunk 8 above)
Comment (Randolph-zeng) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:570]: Just wanna quickly suggest a candidate version of this for loop to avoid the shape doubling bug mentioned above: 
```
        for i, t in enumerate(self.progress_bar(timesteps)):
            if t < t_last:
                # expand the latents if we are doing classifier free guidance
                latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
                latent_model_input = self.scheduler.scale_model_input(latent_model_input, t)

                # predict the noise residual
                noise_pred = self.unet(latent_model_input, t, encoder_hidden_states=text_embeddings).sample

                # perform guidance
                if do_classifier_free_guidance:
                    noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
                    noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)

                # compute the previous noisy sample x_t -> x_t-1
                latents = self.scheduler.step(noise_pred, t, latents, init_latents_orig, mask, generator).prev_sample

                # call the callback, if provided
                if callback is not None and i % callback_steps == 0:
                    callback(i, t, latents)
            else:
                # compute the reverse: x_t-1 -> x_t
                latents = self.scheduler.undo_step(latents, t_last, generator)
                
            t_last = t

```

**Comment by anton-l** [comment]
Thanks for catching the issues @Randolph-zeng!

@nathanielherman let me know if you don't have bandwidth this week, I'd be happy to help getting the PR ready for merging :)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py (hunk 8 above)
Comment (nathanielherman) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:572]: Ah I see, I think I can just put the `if t >= t_last` 2 lines earlier before the `latent_model_input =` to achieve the same effect.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py (hunk 7 above)
Comment (nathanielherman) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:414]: ah true, I updated to remove it entirely since I don't actually use it anymore (repaint just initializes the latents to random noise)

**Comment by nathanielherman** [comment]
Hey! I did most of the updates and will look at adding an integration test. On that note, AFAICT from here https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint_legacy.py#L352, the code is actually loading the non-legacy pipeline rather than the legacy one? I'm a bit confused how that's not breaking it though, since it's initializing it with a non-inpainting model ("CompVis/stable-diffusion-v1-4")

**Comment by Randolph-zeng** [comment]
Just curious, am I the only one that experienced the DDIM degradation here ? When I use the code in this PR I noticed that the DDIM almost failed completely in producing any meaningful impaint image that corresponds to the prompt. 
@nathanielherman Are you troubled by this same issue https://github.com/huggingface/diffusers/issues/1602 or does it work fine with you ?  Thanks a lot if you can share some insight : )

**Comment by nathanielherman** [comment]
@Randolph-zeng hmm I'm confused by the linked issue, do you only get the issue for CFG outside of 6-7, or for any CFG? I only really use the default CFG of 7.5 but for that I get pretty reasonable outputs. (Though I wouldn't say repaint is obviously better results than default inpaint_legacy.)

**Comment by nathanielherman** [comment]
@anton-l bump on my question for the unit test, I just want to make sure I'm understanding correctly before I add my own unit test

**Comment by nathanielherman** [comment]
[red_cat_sitting_on_a_park_bench_repaint.npy.zip](https://github.com/huggingface/diffusers/files/10212497/red_cat_sitting_on_a_park_bench_repaint.npy.zip)
red_cat_sitting_on_a_park_bench

**Comment by nathanielherman** [comment]
@anton-l makes sense! I've added the test and attached the npy file as a comment on this PR — IIUC from the docs, someone would need to upload that npy file and then I can update the test to download it from the url?

**Comment by nathanielherman** [comment]
@anton-l perfect, it should be good for final review now!

**Comment by patrickvonplaten** [comment]
Gently ping @anton-l for a final review

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 9]
```diff
@@ -0,0 +1,572 @@
...
+            DPMSolverMultistepScheduler,
+            RePaintScheduler,
+        ],
+        safety_checker: StableDiffusionSafetyChecker,
+        feature_extractor: CLIPFeatureExtractor,
+        requires_safety_checker: bool = True,
+    ):
+        super().__init__()
+
+        if hasattr(scheduler.config, "steps_offset") and scheduler.config.steps_offset != 1:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:121]: Can we remove all those deprecation messages? We should not add new models with deprecation messages :-)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+                " in future versions. If you have downloaded this checkpoint from the Hugging Face Hub,"
+                " it would be very nice if you could open a Pull request for the `scheduler/scheduler_config.json`"
+                " file"
+            )
+            deprecate("steps_offset!=1", "1.0.0", deprecation_message, standard_warn=False)
+            new_config = dict(scheduler.config)
+            new_config["steps_offset"] = 1
+            scheduler._internal_dict = FrozenDict(new_config)
+
+        if hasattr(scheduler.config, "clip_sample") and scheduler.config.clip_sample is True:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:135]: remove

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py [hunk 11]
```diff
@@ -0,0 +1,572 @@
...
+                "Make sure to define a feature extractor when loading {self.__class__} if you want to use the safety"
+                " checker. If you do not want to use the safety checker, you can pass `'safety_checker=None'` instead."
+            )
+
+        is_unet_version_less_0_9_0 = hasattr(unet.config, "_diffusers_version") and version.parse(
+            version.parse(unet.config._diffusers_version).base_version
+        ) < version.parse("0.9.0.dev0")
+        is_unet_sample_size_less_64 = hasattr(unet.config, "sample_size") and unet.config.sample_size < 64
+        if is_unet_version_less_0_9_0 and is_unet_sample_size_less_64:
+            deprecation_message = (
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_repaint.py:169]: remove

**Comment by patrickvonplaten** [comment]
Thanks for the nice PR @nathanielherman!

Three things from my side:
- 1. Could we also add fast tests here similar to: https://github.com/huggingface/diffusers/blob/f17fae641c7c268e60c479396891ea97a6b29aae/tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint.py#L43
- 2. Could we also add docs as explained here: https://github.com/huggingface/diffusers/tree/main/docs
- 3. Could we remove all the deprecation messages?

Thanks!

File: tests/pipelines/stable_diffusion/test_stable_diffusion_repaint.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+import gc
+import unittest
+
+import numpy as np
+import torch
+
+from diffusers import RePaintScheduler, StableDiffusionRepaintPipeline
+from diffusers.utils import load_image, slow, torch_device
+from diffusers.utils.testing_utils import load_numpy, require_torch_gpu
```
Comment (anton-l) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_repaint.py:24]: As Patrick mentioned above, most of the models are getting covered by common tests that check API compatibility etc. 
What we need here is just a test class similar to `RepaintPipelineFastTests`: https://github.com/huggingface/diffusers/blob/f17fae641c7c268e60c479396891ea97a6b29aae/tests/pipelines/repaint/test_repaint.py#L31
with `pipeline_class = StableDiffusionRepaintPipeline` and slightly adapted `get_dummy_components()` and `get_dummy_inputs()` which you can probably borrow without many changes from `StableDiffusionInpaintPipelineFastTests`: https://github.com/huggingface/diffusers/blob/f17fae641c7c268e60c479396891ea97a6b29aae/tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint.py#L46

Comment (anton-l) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_repaint.py:24]: These added tests will probably uncover some missing pieces in the pipeline, so feel free to ping us if something is tough to fix! :)

**Comment by vlordier** [comment]
@anton-l is there something left we can do to merge this pipeline ?

**Comment by anton-l** [comment]
@vlordier the TODO is mostly just to update the tests as per the comments above (and fix any API issues uncovered by the common tests), and resolve the merge conflicts.
